### ISSUE TITLE
Quantize model: does not have MinMax information, and is not a constant array when quantize the concat op layer

### ISSUE THREAD

**Issue by 1icas** [tensorflow/tensorflow#21725]
tensorflow verison: 1.10
problem:I will describle this problem in two cases(one can quantize succeed, the other can not success).
tf.concat(
    values,
    axis,
    name='concat'
). This's the concat api.

1.if i use concat like this： tf.concat(values=['t1', 't2'], axis=3). My model can quantize succeed.
2.if i use concat like this： tf.concat(values=['t1'], axis=3).some errors occured: concat does not have MinMax information, and is not a constant array. Cannot proceed with quantization.
t1 and t2 are the layer name.

I quantize my model like this:
bazel-bin/tensorflow/contrib/lite/toco/toco \
--input_file=/home/admin_pc/model_test/output_qt1.pb \
--output_file=/home/admin_pc/model_test/mobilenet_qg.tflite \
--input_fromat=TENSORFLOW_GRAPHDEF \
--output_format=TFLITE \
--inference_type=QUANTIZED_UINT8 \
--input_array=image \
--output_array=Pose/concat_stage7 \
--input_shape=1,224,224,3 \
--change_concat_input_ranges=false

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by asimshankar** [comment]
@suharshs : Mind taking a look?

**Comment by suharshs** [comment]
Hi,

What model are you trying to quantize? And can you provide what the input nodes to your concat operation are in the success case and the failure case? And can you paste the exact error message from TOCO?

Thanks!

**Comment by 1icas** [comment]
@suharshs , i can provide the .pb file. do you need?

**Comment by 1icas** [comment]
@suharshs ,hi. can you tell me this's tensorflow bug or my source bug ?  If my source bug,i will check my model and source code soon. thanks

**Comment by sepulm01** [comment]
Hi, I have the same issue "Array TFLite_Detection_PostProcess does not have MinMax information, and is not a constant array. Cannot proceed with quantization."
My tensorflow version is 1.10.1 with GPU. I've training my data with ssd_mobilenet_v1_0.75_depth_quantized_300x300_pets_sync.config in Google Machine Learning (TPU). 
 The exported method that I use is object_detection/export_tflite_ssd_graph.py, I did TOCO with  "pet breeds dataset" example and TOCO works well. My problem is when I try to use TOCO with my own data set. I did the very same process for both data set (my own and TF example). The frozen graph with my own data set woks well in my computer (object_detection/export_inference_graph.py). 

TOCO command:
bazel run -c opt tensorflow/contrib/lite/toco:toco -- \
--input_file=$OUTPUT_DIR/tflite_graph.pb \
--output_file=$OUTPUT_DIR/detect.tflite \
--input_shapes=1,300,300,3 \
--input_arrays=normalized_input_image_tensor \
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3'  \
--inference_type=QUANTIZED_UINT8 \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops

best regards.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sepulm01** [comment]
yep ! I solve with apply fake Quatized

--default_ranges_min=0 \
--default_ranges_max=6 \

I hope it will be usefull to other people :)

**Comment by suharshs** [comment]
Hi @1icas, 

The issue is that the contrib/quantize rewriter is not very robust to any arbitrary model yet. In particular the concats in your network need quantization information for TOCO and aren't supported out of the box. This can be resolved by either manually adding a FakeQuantWithMinMaxVars node after the concats, see how the rewriter does it: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/quantize/python/quantize.py#L475

That being said this can be complicated and very error prone, if you goal is to just get a smaller and faster model, I recommend trying the --post_training_quantize flag to tflite_convert. With that you keep the inference_type=FLOAT and pass a floating point version of your model (no need to call the contrib/quantize tool). That may provide sufficient speedup for your use case. If it doesn't provide sufficient speedup (or if it has an error, please file another issue!!) then we can consider this full quantization.

**Comment by MichaelHunson** [comment]
> yep ! I solve with apply fake Quatized
> 
> --default_ranges_min=0 
> --default_ranges_max=6 \
> 
> I hope it will be usefull to other people :)



> yep ! I solve with apply fake Quatized
> 
> --default_ranges_min=0 
> --default_ranges_max=6 \
> 
> I hope it will be usefull to other people :)

Hello @sepulm01 , thanks for your suggestion to solve this issue. I have encountered the same problem and solved it with your way. However, I am wondering how to set the --default_ranges_min and --default_ranges_max values? The 0 and 6 are obtained via some calculation or something else? 

Thanks,
Yuhang

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Vincent630** [comment]
thanks for your suggestion to solve this issue. I have encountered the same problem and solved it with your way. but when i evalute the result of converted tflite model,i get a zerodivisionerror:float (num_correct)/float(cnt_total),did you get this issues?if you have any idea about this please let me know,thaks a lot

**Comment by hahadashi** [comment]
> Hi,
> 
> What model are you trying to quantize? And can you provide what the input nodes to your concat operation are in the success case and the failure case? And can you paste the exact error message from TOCO?
> 
> Thanks!

hi , i meet some some problem when i want to quatize my model.(cnn+max_pool+bn+relu+fc). but it returns a error "FusedBatchNorm_mul_0, which is an input to the Add operator producing the output array kws_model/KWS_Model/tower_0/CNN_V1/Relu, is lacking min/max data, which is necessary for quantization. Either target a non-quantized output format, or change the input graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you do not care about the accuracy of results." i tried to add default value can solve the problem, but i know it the best solution. 

or i also tried to remove the bn layer , there is not the error , but it return a new error message "prediction does not have MinMax information, and is not a constant array. Cannot proceed with quantization." the prediction is the last layer of my network,
(final_fc_weights = tf.get_variable("final_fc_weights", [cfg.fc2_size, cfg.n_class], initializer=initializer)
174     final_fc_bias = tf.get_variable("final_fc_bias", [cfg.n_class], initializer=initializer)
176     final_fc = tf.add(tf.matmul(fc2, final_fc_weights), final_fc_bias, name="prediction")
177     return final_fc
)
hope you give me some suggestions, thanks.

**Comment by suharshs** [comment]
Based on feedback that the contrib/quantize quantization-aware training tool is a bit brittle and hard to use on some model architectures, we have released a [post-training integer quantization tool](https://medium.com/tensorflow/tensorflow-model-optimization-toolkit-post-training-integer-quantization-b4964a1ea9ba), that requires a small calibration dataset. Please take a look at the [tutorial](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tutorials/post_training_integer_quant.ipynb) and give it a try, it should work much better! And let us know if you run into any issues.

Closing this issue, since we are rethinking and working on an api to replace contrib/quantize quantization-aware-training (although post-training quantization above should be sufficient for the majority of use cases).

Thanks!
-Suharsh

**Comment by bayesian-mind** [comment]
> Hi @1icas,
> 
> The issue is that the contrib/quantize rewriter is not very robust to any arbitrary model yet. In particular the concats in your network need quantization information for TOCO and aren't supported out of the box. This can be resolved by either manually adding a FakeQuantWithMinMaxVars node after the concats, see how the rewriter does it: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/quantize/python/quantize.py#L475
> 
> That being said this can be complicated and very error prone, if you goal is to just get a smaller and faster model, I recommend trying the --post_training_quantize flag to tflite_convert. With that you keep the inference_type=FLOAT and pass a floating point version of your model (no need to call the contrib/quantize tool). That may provide sufficient speedup for your use case. If it doesn't provide sufficient speedup (or if it has an error, please file another issue!!) then we can consider this full quantization.

@suharshs i tried clicking on the link above and I was redirected to a non existent page. I am having a similar issue where I was successfully able to quantize and produce a tflite graph but I noticed that minmaxvars node was missing after the concatV2 node. Can you please provide with the link agagin or direct me to something that might help?

**Comment by suharshs** [comment]
Here is an updated link https://github.com/tensorflow/tensorflow/tree/r1.8/tensorflow/contrib/quantize

We have a new Keras 2.0 QAT api that is more configurable, please take a look here: https://blog.tensorflow.org/2020/04/quantization-aware-training-with-tensorflow-model-optimization-toolkit.html

**Comment by dtlam26** [comment]
@suharshs do you have any example of manually adding the quatize in contrib to the graph?
It is a little hard to verify the way of excecuting it
